syntax = "proto3";

service RaftService {
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);

    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);

    rpc GetLeader(GetLeaderRequest) returns (GetLeaderResponse);

    rpc SetVal(SetValRequest) returns (SetValResponse);

    rpc GetVal(GetValRequest) returns (GetValResponse);
}

message RequestVoteRequest {
    // CANDIDATE term and id
    int32 candidateTerm = 1;
    int32 candidateId = 2;
    int32 lastLogIndex = 3;
    int32 lastLogTerm = 4;
}

message RequestVoteResponse {
    int32 term = 1;
    
    // results of the vote
    bool result = 2;
    float prevLeaseTime = 3;
}

message Entry {
    int32 term = 1;
    repeated string update = 2;

}

message AppendEntriesRequest {
    // LEADER term and id
    int32 leaderTerm = 1;
    int32 leaderId = 2;
    int32 prevLogIndex = 3;
    int32 prevLogTerm = 4;
    repeated Entry entries = 5;
    int32 leaderCommit= 6;
    float remainingLeaseTime = 7;
}

message AppendEntriesResponse {
    int32 term = 1;
    bool success = 2;
}

message GetLeaderRequest {}

message GetLeaderResponse {
    // LEADER id and address 
    int32 leaderId = 1;
    string address = 2;
}

message SetValRequest {
    string key = 1;
    string value = 2;
}

message SetValResponse {
    bool success = 1;
    int32 leaderID = 2;
}

message GetValRequest {
    string key = 1;
}

message GetValResponse {
    bool success = 1;
    string value = 2;
    int32 leaderID = 3;    
}